### PR TITLE
feat(vmop): implement conditions, use new reconciler style

### PR DESCRIPTION
**Pull request by diafour**
## Description

- Refactor vmop controller to use new style of reconcilers.
- Add conditions to the VMOP status.
- Remove old-style failureMessage/failureReason fields.

Behaviour changes:
- Do not wait for VM, set VMOP to failed state if the specified VM is absent.
- Do not wait for other VMOPs to complete, set VMOP to failed state if other VMOP is in progress.
- Do not add protection finalizer on VM.
- Add label to VMOP with vm name.
- Add type to additionalPrinterColumns

image

## Why do we need it, and what problem does it solve?

It is the last refactoring that brings Conditions to VirtualMachineOperation resources.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: crds/virtualmachineoperations.yaml [hunk 1]
```diff
@@ -64,8 +64,41 @@ spec:
...
+                      status:
+                        description: Status of the condition, one of True, False, Unknown.
+                        type: string
+                        enum: ["True", "False", "Unknown"]
+                      type:
+                        description: Type of condition.
+                        type: string
+                    required:
+                      - status
+                      - type
```
Comment (yaroslavborbat) [inline crds/virtualmachineoperations.yaml:101]: The reason must be a required field

File: images/virtualization-artifact/pkg/controller/vmop/internal/state/state.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	VirtualMachine(ctx context.Context) (*virtv2.VirtualMachine, error)
+}
+
+func New(c client.Client, vmop *service.Resource[*virtv2.VirtualMachineOperation, virtv2.VirtualMachineOperationStatus]) VMOperationState {
+	return &state{client: c, vmop: vmop}
+}
+
+type state struct {
+	client client.Client
+	mu     sync.RWMutex
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vmop/internal/state/state.go:43]: mutex is not used

File: images/virtualization-artifact/pkg/controller/vmop/internal/util.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	if vmop == nil {
+		return false
+	}
+
+	if vmop.Status.Phase == virtv2.VMOPPhaseCompleted || vmop.Status.Phase == virtv2.VMOPPhaseFailed {
+		return true
+	}
+
+	return false
+}
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vmop/internal/util.go:31]: consider adding this function as a method to a resource VMOP.

File: images/virtualization-artifact/pkg/controller/service/vm_operation.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+func (s VMOperationService) getKVVM(ctx context.Context, vmNamespace, vmName string) (*virtv1.VirtualMachine, error) {
+	kvvm := virtv1.VirtualMachine{}
+	key := types.NamespacedName{Name: vmName, Namespace: vmNamespace}
+	err := s.client.Get(ctx, key, &kvvm)
+	if err != nil {
+		return nil, err
+	}
+	return &kvvm, nil
+}
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/service/vm_operation.go:51]: It seems to me that if the function can return not found error, then we should not check the object for nil.

So I see this method like this:
```go
kvvm := &virtv1.VirtualMachine{}
key := types.NamespacedName{Name: vmName, Namespace: vmNamespace}
err := s.client.Get(ctx, key, kvvm)
return kvvm, err
```
or
```
key := types.NamespacedName{Name: vmName, Namespace: vmNamespace}
return helper.FetchObject(ctx,
		key,
		client,
		&virtv1.VirtualMachine{})
```

File: images/virtualization-artifact/pkg/controller/service/vm_operation.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+		return vmopcondition.StartInProgress
+	case virtv2.VMOPTypeStop:
+		return vmopcondition.StopInProgress
+	case virtv2.VMOPTypeRestart:
+		return vmopcondition.RestartInProgress
+	}
+	return ""
+}
+
+func (s VMOperationService) IsComplete(vmop *virtv2.VirtualMachineOperation, vm *virtv2.VirtualMachine) bool {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/service/vm_operation.go:194]: We already have method IsCompleted. 
Consider deleting one of them. Or reuse

File: images/virtualization-artifact/pkg/controller/service/vm_operation.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	vmPhase := vm.Status.Phase
+
+	switch vmopType {
+	case virtv2.VMOPTypeRestart, virtv2.VMOPTypeStart:
+		return vmPhase == virtv2.MachineRunning
+	case virtv2.VMOPTypeStop:
+		return vmPhase == virtv2.MachineStopped
+	default:
+		return false
+	}
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/service/vm_operation.go:209]:
```suggestion
	return s.IsCompleted(vmopType, vmPhase)
```

File: images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+	// Initialize new VMOP resource: set phase to Pending and all conditions to Unknown.
+	if changed.Status.Phase == "" {
+		// TODO add label with vm name.
+		changed.Status.Phase = virtv2.VMOPPhasePending
+		// Add all conditions in unknown state.
+		for _, condType := range lifeCycleConditions {
+			service.SetCondition(metav1.Condition{
+				Type:   condType,
+				Status: metav1.ConditionUnknown,
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go:95]: Add Unknown Reason

File: images/virtualization-artifact/pkg/controller/vmop/internal/protection.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+func (h ProtectionHandler) Handle(ctx context.Context, s state.VMOperationState) (reconcile.Result, error) {
+	if s.VirtualMachineOperation() == nil {
+		return reconcile.Result{}, nil
+	}
+
+	changed := s.VirtualMachineOperation().Changed()
+	log := h.logger.With("name", changed.GetName(), "namespace", changed.GetNamespace())
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vmop/internal/protection.go:53]: use logger.FromContext(ctx)

File: images/virtualization-artifact/pkg/controller/vmop/vmop_controller.go [hunk 9]
```diff
@@ -38,41 +40,44 @@ const (
...
+	log = log.With(logger.SlogController(controllerName))
+
+	recorder := mgr.GetEventRecorderFor(controllerName)
+	client := mgr.GetClient()
+	vmopSrv := service.NewVMOperationService(mgr.GetClient())
+
+	handlers := []Handler{
+		internal.NewLifecycleHandler(log, vmopSrv),
+		internal.NewOperationHandler(log, recorder, vmopSrv),
+		internal.NewProtectionHandler(log, client),
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vmop/vmop_controller.go:54]: rm log from handlers

**Review by yaroslavborbat — DISMISSED** [review]

File: images/virtualization-artifact/pkg/controller/vmop/vmop_webhook.go [hunk 10]
```diff
@@ -59,7 +59,7 @@ func (v *Validator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Obj
 	if reflect.DeepEqual(oldVmop.Spec, newVmop.Spec) {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmop/vmop_webhook.go:59]: Why not just compare the generation of different versions?

File: images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+)
+
+const lifecycleHandlerName = "LifecycleHandler"
+
+var lifeCycleConditions = []vmopcondition.Type{
+	vmopcondition.CompletedType,
+	vmopcondition.SignalSentType,
+}
+
+// LifecycleHandler calculate status of the VirtualMachineOperation resource.
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go:40]: calculates

File: api/core/v1alpha2/vmopcondition/condition.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	NotApplicableForRunPolicy CompletedReason = "NotApplicableForRunPolicy"
+
+	// NotApplicableForVMPhase is a CompletedReason indicating that the specified operation type is not appilicable for the virtual machine phase.
+	NotApplicableForVMPhase CompletedReason = "NotApplicableForVMPhase"
+
+	// OtherOperationsAreInProgress is a CompletedReason indicating that there are other operations in progress.
+	OtherOperationsAreInProgress CompletedReason = "OtherOperationsAreInProgress"
+
+	// // WaitForOtherOperations is a CompletedReason indicating that there are other operations in progress.
+	// WaitForOtherOperations CompletedReason = "WaitForOtherOperations"
```
Comment (Isteb4k) [inline api/core/v1alpha2/vmopcondition/condition.go:48]: Unused reason

File: api/core/v1alpha2/vmopcondition/condition.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+	// CompletedType is a type for condition that indicates operation is complete.
+	CompletedType Type = "Completed"
+
+	// SignalSentType is a type for condition that indicates operation signal has been sent.
+	SignalSentType Type = "SignalHasBeenSent"
+)
+
+type (
+	// CompletedReason represents specific reasons for the 'Completed' condition type.
+	CompletedReason = string
```
Comment (Isteb4k) [inline api/core/v1alpha2/vmopcondition/condition.go:31]: What about reasons for SignalSentType?

File: images/virtualization-artifact/pkg/controller/vmop/vmop_controller.go (hunk 9 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/vmop_controller.go:54]: Done

File: images/virtualization-artifact/pkg/controller/vmop/internal/protection.go (hunk 8 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/protection.go:53]: Done

File: images/virtualization-artifact/pkg/controller/service/vm_operation.go (hunk 5 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/service/vm_operation.go:194]: Forget delete IsCompleted, done.

File: images/virtualization-artifact/pkg/controller/vmop/internal/protection.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+	vm, _ := s.VirtualMachine(ctx)
+
+	// The only case when we need finalizer on VirtualMachineOperation is when operation is in progress.
+	if changed.DeletionTimestamp == nil && changed.Status.Phase == virtv2.VMOPPhaseInProgress {
+		log.Debug("Add protection finalizer for the InProgress phase")
+		controllerutil.AddFinalizer(changed, virtv2.FinalizerVMOPCleanup)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmop/internal/protection.go:60]: We lost the distinction between cleanup and protection finalizers. Are you sure we need cleanup here? It looks like we need protection.

File: images/virtualization-artifact/pkg/controller/vmop/internal/protection.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+	return reconcile.Result{}, nil
+}
+
+func (h ProtectionHandler) Name() string {
+	return protectionHandlerName
+}
+
+// ensureVMFinalizers ensures that VM has a finalizer.
+// TODO refactor with patching.
+func (h ProtectionHandler) ensureVMFinalizers(ctx context.Context, vm *virtv2.VirtualMachine) error {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmop/internal/protection.go:82]: Use protection service here (.protection.AddProtection)

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vmop/internal/util.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+
+func IsFinalState(vmop *virtv2.VirtualMachineOperation) bool {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmop/internal/util.go:21]: Please, avoid creating new util functions

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: crds/virtualmachineoperations.yaml [hunk 17]
```diff
@@ -64,8 +64,41 @@ spec:
...
                 phase:
                   type: string
                   description: |
```
Comment (Isteb4k) [inline crds/virtualmachineoperations.yaml:104]: Remove "See `.status.failureReason`." from Phase description

File: images/virtualization-artifact/pkg/controller/vmop/internal/util.go (hunk 3 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/util.go:31]: Moved to vmop service

File: images/virtualization-artifact/pkg/controller/vmop/internal/state/state.go (hunk 2 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/state/state.go:43]: Removed

File: crds/virtualmachineoperations.yaml (hunk 17 above)
Comment (Isteb4k) [reply crds/virtualmachineoperations.yaml:104]: And same for ru version of doc

File: crds/virtualmachineoperations.yaml (hunk 1 above)
Comment (diafour) [reply crds/virtualmachineoperations.yaml:101]: Only VMClass has additional fields in required:
    "lastTransitionTime",
    "message",
    "reason",

File: images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	// Fail if VirtualMachineOperation is not applicable for VM phase.
+	if !h.vmopSrv.IsApplicableForVMPhase(changed, vm.Status.Phase) {
+		changed.Status.Phase = virtv2.VMOPPhaseFailed
+		service.SetCondition(metav1.Condition{
+			Type:    vmopcondition.CompletedType,
+			Reason:  vmopcondition.NotApplicableForVMPhase,
+			Status:  metav1.ConditionFalse,
+			Message: fmt.Sprintf("Operation type %s is not applicable for vm phase %s", changed.Spec.Type, vm.Status.Phase),
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go:163]: Avoid using abbreviations for user messages

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go (hunk 7 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go:95]: Done

File: api/core/v1alpha2/vmopcondition/condition.go (hunk 13 above)
Comment (diafour) [reply api/core/v1alpha2/vmopcondition/condition.go:31]: Added

File: api/core/v1alpha2/vmopcondition/condition.go (hunk 12 above)
Comment (diafour) [reply api/core/v1alpha2/vmopcondition/condition.go:48]: Removed

File: images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go (hunk 11 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go:40]: Done

File: images/virtualization-artifact/pkg/controller/vmop/internal/util.go (hunk 16 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/util.go:21]: Moved to service

File: images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go (hunk 18 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go:163]: Rewritten

			Message: fmt.Sprintf("Operation type %s is not applicable for VirtualMachine with runPolicy %s", changed.Spec.Type, vm.Spec.RunPolicy),
			Message: fmt.Sprintf("Operation type %s is not applicable for VirtualMachine in phase %s", changed.Spec.Type, vm.Status.Phase),

File: crds/virtualmachineoperations.yaml (hunk 17 above)
Comment (diafour) [reply crds/virtualmachineoperations.yaml:104]: Done

File: images/virtualization-artifact/pkg/controller/vmop/internal/protection.go (hunk 14 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/protection.go:60]: Renamed to deletion. Do not add protection finalizers on VMs any more.

File: images/virtualization-artifact/pkg/controller/vmop/internal/protection.go (hunk 15 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/protection.go:82]: Not needed

File: images/virtualization-artifact/pkg/controller/service/vm_operation.go (hunk 4 above)
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/service/vm_operation.go:51]: Changed to FetchObject

File: crds/virtualmachineoperations.yaml (hunk 1 above)
Comment (diafour) [reply crds/virtualmachineoperations.yaml]: Add required fields, use Builder and conditions.SetCondition

File: api/core/v1alpha2/vmopcondition/condition.go [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+const (
+	// TypeCompleted is a type for condition that indicates operation is complete.
+	TypeCompleted Type = "Completed"
+
+	// SignalSentType is a type for condition that indicates operation signal has been sent.
+	SignalSentType Type = "SignalSent"
+)
+
+// Reason represents specific reasons for the 'Completed' condition type.
```
[the remark this answers is not in the document]
Comment (diafour) [reply api/core/v1alpha2/vmopcondition/condition.go]: Splitted

File: api/core/v1alpha2/vmopcondition/condition.go [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+	// ReasonSignalSentError is a Reason indicating an error occurred while sending powerstate signal to the VM.
+	ReasonSignalSentError Reason = "ReasonSignalSentError"
+
+	// ReasonSignalSentSuccess is a Reason indicating that signal is sent to the VM.
+	ReasonSignalSentSuccess Reason = "ReasonSignalSentSuccess"
+
+	// ReasonVirtualMachineNotFound is a Reason indicating that the specified virtual machine is absent.
+	ReasonVirtualMachineNotFound Reason = "ReasonVirtualMachineNotFound"
+
```
[the remark this answers is not in the document]
Comment (diafour) [reply api/core/v1alpha2/vmopcondition/condition.go:51]: Done

File: images/virtualization-artifact/pkg/controller/vmop/internal/deletion.go [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+	if s.VirtualMachineOperation() == nil {
+		return reconcile.Result{}, nil
+	}
+
+	changed := s.VirtualMachineOperation().Changed()
+
+	// The only case when we need finalizer on VirtualMachineOperation is when operation is in progress.
+	if changed.DeletionTimestamp == nil && changed.Status.Phase == virtv2.VMOPPhaseInProgress {
+		log.Debug("Add cleanup finalizer while in the InProgress phase")
+		controllerutil.AddFinalizer(changed, virtv2.FinalizerVMOPCleanup)
```
[the remark this answers is not in the document]
Comment (diafour) [reply images/virtualization-artifact/pkg/controller/vmop/internal/deletion.go:57]: Removed

**Review by Isteb4k — APPROVED** [review]

**Review by yaroslavborbat — APPROVED** [review]

**Pull request merged**